### PR TITLE
Report gradient norm and clipping statistics

### PR DESCRIPTION
**Pull request by stephenroller**
**Patch description**
Generally useful metrics to have when training a model (borrowing from fairseq). `gnorm` reports the 2-norm of the parameter weights (treated as a vector). `clip` reports the fraction of updates whose gradient was clipped.

Currently only reported when actually using `--gradient-clip` but we could change that.

**Testing steps**
Been using extensively for a few weeks. Needs another behavior when not using clipping.

### CODE REVIEW

File: parlai/core/torch_agent.py [hunk 1]
```diff
@@ -215,6 +215,13 @@ def reset(self):
         self.history_strings = []
         self.history_vecs = []
 
+    def reset_metrics(self):
```
Comment (emilydinan) [inline parlai/core/torch_agent.py]: why is this in the History object?????? super().reset_metrics() here isn't going to do anything?

File: parlai/core/torch_agent.py [hunk 2]
```diff
@@ -941,6 +948,11 @@ def report(self):
             current_lr = round_sigfigs(self.optimizer.param_groups[0]['lr'], 4)
             metrics['lr'] = round_sigfigs(current_lr, 4)
         metrics['num_updates'] = self._number_training_updates
+
+        steps = max(1, self.metrics['updates'])  # prevent divide by zero
```
Comment (emilydinan) [inline parlai/core/torch_agent.py]: again,  i don't think this is  in the right place. the history objects doesn't have a metrics attribute

Comment (stephenroller) [reply parlai/core/torch_agent.py]: O_o what in teh world!

**Comment by stephenroller** [comment]
Alright, ready for another review. no rush.

File: parlai/core/torch_agent.py [hunk 3]
```diff
@@ -1101,6 +1116,14 @@ def share(self):
         Subclasses will likely want to share their model as well.
         """
         shared = super().share()
+
+        if self.opt.get('numthreads', 1) > 1 and isinstance(self.metrics, dict):
+            torch.set_num_threads(1)
```
Comment (emilydinan) [inline parlai/core/torch_agent.py]: what does this call do? looks like we weren't calling this in TGA before if `self.opt['numthreads'] > 1`

File: parlai/core/torch_agent.py [hunk 4]
```diff
@@ -941,6 +951,11 @@ def report(self):
             current_lr = round_sigfigs(self.optimizer.param_groups[0]['lr'], 4)
             metrics['lr'] = round_sigfigs(current_lr, 4)
         metrics['num_updates'] = self._number_training_updates
+
+        steps = max(1, self.metrics['updates'])  # prevent divide by zero
+        if steps > 0 and self.opt.get('gradient_clip', -1) > 0:
+            metrics['gnorm'] = round_sigfigs(self.metrics['gnorm'] / steps, 4)
+            metrics['clip'] = round_sigfigs(self.metrics['clip'] / steps, 2)
```
[the remark this answers is not in the document]
Comment (stephenroller) [reply parlai/core/torch_agent.py]: 4 is gratuitous

File: parlai/core/torch_agent.py (hunk 3 above)
Comment (stephenroller) [reply parlai/core/torch_agent.py]: It prevents torch from parallelizing the matrix multiplication across cores (resulting in hogwild stomping each other).

We actually already do it in `__init__` so I can drop this line.

**Comment by stephenroller** [comment]
Comments addressed
